shell/platform/android/io/flutter/embedding/engine/dynamicfeatures/PlayStoreDynamicFeatureManager.java

```diff
@@ -0,0 +1,336 @@
+    // context, neither parameter is used here. Assets are stored in
+    // the apk's `assets` directory allowing them to be accessed by
+    // Android's AssetManager directly.
+    try {
+      context = context.createPackageContext(context.getPackageName(), 0);
+
+      AssetManager assetManager = context.getAssets();
+      flutterJNI.updateAssetManager(
+          assetManager,
+          // TODO(garyq): Made the "flutter_assets" directory dynamic based off of DartEntryPoint.
+          "flutter_assets");
+    } catch (NameNotFoundException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  public void loadDartLibrary(int loadingUnitId, String moduleName) {
+    if (!verifyJNI()) {
+      return;
+    }
+    // Loading unit must be specified and valid to load a dart library.
+    if (loadingUnitId < 0) {
+      return;
+    }
+
+    // This matches/depends on dart's loading unit naming convention, which we use unchanged.
+    String aotSharedLibraryName = "app.so-" + loadingUnitId + ".part.so";
+
+    // Possible values: armeabi, armeabi-v7a, arm64-v8a, x86, x86_64, mips, mips64
+    String abi;
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
+      abi = Build.SUPPORTED_ABIS[0];
+    } else {
+      abi = Build.CPU_ABI;
+    }
+    String pathAbi = abi.replace("-", "_"); // abis are represented with underscores in paths.
+
+    // TODO(garyq): Optimize this apk/file discovery process to use less i/o and be more
+    // performant and robust.
+
+    // Search directly in APKs first
+    List<String> apkPaths = new ArrayList();
+    // If not found in APKs, we check in extracted native libs for the lib directly.
+    List<String> soPaths = new ArrayList();
+    Queue<File> searchFiles = new LinkedList();
+    searchFiles.add(context.getFilesDir());
+    while (!searchFiles.isEmpty()) {
+      File file = searchFiles.remove();
+      if (file != null && file.isDirectory()) {
+        for (File f : file.listFiles()) {
+          searchFiles.add(f);
+        }
+        continue;
+      }
+      String name = file.getName();
+      if (name.endsWith(".apk") && name.startsWith(moduleName) && name.contains(pathAbi)) {
+        apkPaths.add(file.getAbsolutePath());
+        continue;
+      }
+      if (name.equals(aotSharedLibraryName)) {
+        soPaths.add(file.getAbsolutePath());
+      }
+    }
+
+    List<String> searchPaths = new ArrayList();
+    for (String path : apkPaths) {
+      searchPaths.add(path + "!lib/" + abi + "/" + aotSharedLibraryName);
+    }
+    for (String path : soPaths) {
+      searchPaths.add(path);
+    }
+
+    flutterJNI.loadDartDeferredLibrary(
+        loadingUnitId, searchPaths.toArray(new String[apkPaths.size()]));
+  }
+
+  public void uninstallFeature(int loadingUnitId, String moduleName) {
+    // TODO(garyq): support uninstalling.
+  }
+
+  public void destroy() {
+    splitInstallManager.unregisterListener(listener);
```

Comment:
while you're here, might as well remove the JNI reference. i.e. during the destruction sequence, there is no guarantee that there won't be a race where a load asset platform channel call lands right as some items are being destroyed.

Reply:
You didn't address this one.